document/chapters/related_work.tex

```diff
@@ -1,3 +1,83 @@
 % !TEX root = ../main.tex
-\chapter{Related Work}
-\label{chapter:related_work}
+\chapter{Related Work} \label{chapter:related_work}
+
+\section{The theory of dependent types}
+
+\section{Dependently typed programming languages}
+
+Given the correspondence between dependent type systems and propositional logic.
+The first programming languages with dependent types were proof assistants, such
+as Coq and Agda, \cite{agda} which can be used to proof theorems automatically.
+However, such languages are not used outside the academic community nor to write
+user or system oriented applications.
+
+Haskell, being both a general purpose programming language and the de facto
+language to explore and study new type systems in the academy, has provided the
+foundation to study several formulations of dependent types in the form of new
+languages, such as Agda, Idris \cite{idris} and Cayenne, \cite{cayenne} or by
+adding dependent types directly to the language. This last part is particularly
+relevant to this work, given that Rust's nor Haskell's type system and compiler
+were written to have dependent types.
+
+C++ is the language that is most often compared to Rust because both languages
+were designed to write performance oriented applications. Templates, which are
+C++ mechanism to achieve parametric polymorphism, allow not only types as
+parameters but also constant values. \cite{templates} This idea is pretty
+similar to the one being implemented for Rust in the RFC 2000. Even then, manual
+memory allocation makes C++ a language far from being memory safe.  Where as
+Rust is a completely memory safe language.    
+
+\subsection{Dependent Haskell} 
+
+The Glasgow Haskell compiler has certain features which makes Haskell almost a
+dependently typed language. For example, datatype promotion allows to promote
+algebraic data types to be promoted as kinds, and each term of such types to be 
+promoted as a type itself. With these promotions, traditional polymorphism and
+type families can be seen as a form of dependent types with some restrictions.
+
+The addition of fully dependent types to Haskell's type
+system has been studied deeply by Gundry \cite{gundry} and subsequently by
```

Comment:
```suggestion
system has been studied deeply by~\citet{gundry} and subsequently by
```

Reply:
Using natbib, you can add the names to the authors of a publication simply writing \citet{} instead of \cite